### ISSUE TITLE
How to implement the unconcat version?

### ISSUE THREAD

**Issue by liziyun2013** [DonkeyShot21/UNO#11]
Hi Enrico,

Thanks for your work and clean code!
From your paper, I find that acc increases dramatically after concatenating label's and unlabel's logistics.
And I really want to see how to implement the unconcat version.
Looking forwards to your reply.

**Comment by DonkeyShot21** [comment]
Hi,

Thanks for your interest in our work.

I don't remember exactly how I implemented it but for sure you will need to:
1) do not concat here
https://github.com/DonkeyShot21/UNO/blob/1d7811f387b5855025e11cfadbebc1b72879eba0/main_discover.py#L155
2) the shape of this tensor will also change
https://github.com/DonkeyShot21/UNO/blob/1d7811f387b5855025e11cfadbebc1b72879eba0/main_discover.py#L164
3) change the slicing in this loop
https://github.com/DonkeyShot21/UNO/blob/1d7811f387b5855025e11cfadbebc1b72879eba0/main_discover.py#L167
4) use the swapped prediction only for the unlab samples
https://github.com/DonkeyShot21/UNO/blob/1d7811f387b5855025e11cfadbebc1b72879eba0/main_discover.py#L179
5) here add cross-entropy on the labeled samples, something like `F.cross_entropy(logits_lab, targets_lab)`
https://github.com/DonkeyShot21/UNO/blob/1d7811f387b5855025e11cfadbebc1b72879eba0/main_discover.py#L189

Something like this. It should not be too hard. Remember that the repo now contains code for UNO v2 that performs much better than the original version, so results are gonna be different.

**Comment by liziyun2013** [comment]
what's the main difference between v1 and v2? I note that you fix cross_entropy loss.
And I also have another problem? do you think multiple heads(unlabel heads) help a lot for the performance?

**Comment by liziyun2013** [comment]
Thank you so much for the fast reply! I will implement it following your guides

**Comment by DonkeyShot21** [comment]
The main difference is in the augmentations. I recommend you use the new version. With UNO v2 multiple heads do not help as much as before, so if you want to speed up computation you can decrease the number of heads.

**Comment by JosephKJ** [comment]
Hi Enrico,

> The main difference is in the augmentations.

But, I find that both `transforms.py` are identical in [master](https://github.com/DonkeyShot21/UNO/blob/main/utils/transforms.py) and [v2](https://github.com/DonkeyShot21/UNO/blob/v2/utils/transforms.py) branch. Had the new changes been ported to both the branches alike?

Kindly clarify when you are free. Thanks in advance!

~ Joseph

**Comment by DonkeyShot21** [comment]
The branch has been merged into main.

**Comment by JosephKJ** [comment]
Great! Thanks!
